### PR TITLE
Added piston control java file

### PR DESCRIPTION
**Pull request by SteepAtticStairs**
Added controlling pistons/solenoids in java-libs

### CODE REVIEW

File: subsystems/PistonControl.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+import static edu.wpi.first.wpilibj.DoubleSolenoid.Value.*;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public abstract class PistonControl extends SubsystemBase{
+    private DoubleSolenoid[] doubleSolenoidPistons;
+
+    public PistonControl(int[] pistonIDS) {
+        for(int i = 0; i < pistonIDS.length; i+=2) {  // for the amount of numbers in the pistonIDS array defined in constants
+            doubleSolenoidPistons[i] = new DoubleSolenoid(i, i+1);  // create a new double solenoid
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: I think `pistonIDs` should be a 2D array, so then each row would represent both ports of the double solenoid. So the array would be 2 by the number of solenoids.

File: subsystems/PistonControl.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        }
+
+    }
+
+    public void extend() {  // method to extend the piston(s)
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: Documentation for a method should be in the javadoc standard. Check this out if you don't know how: https://www.tutorialspoint.com/java/java_documentation.htm

File: subsystems/PistonControl.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        for(DoubleSolenoid piston : doubleSolenoidPistons) { // extends the piston for every piston created
+            piston.set(kForward);
+        }
+    }
+
+    public void retract() {  // method to retract the piston(s)
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: same here

**Review by mxgordon — CHANGES_REQUESTED** [review]

**Comment by csokolove** [comment]
So the pistonsSet method may have simplified things at the abstract level, but complicated things at the project level. We can look at it - but may just be best to remove it and go back.

File: subsystems/PistonControl.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+
+import static edu.wpi.first.wpilibj.DoubleSolenoid.Value;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class PistonControl extends SubsystemBase{
+    private DoubleSolenoid[] doubleSolenoidPistons;
+
+    public PistonControl(int[] pistonIDS) {
+        for(int i = 0; i < pistonIDS.length; i+=2) {
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: before this for loop, you should add `doubleSolenoidPistons = new DoubleSolenoid[pistonIDS.length/2];` to initialize the array

File: subsystems/PistonControl.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+            doubleSolenoidPistons[i] = new DoubleSolenoid(i, i+1);
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: This shouldn't be `i` and `i + 1`. It should be `pistonIDs[i]` and `pistonIDs[i + 1]`

File: subsystems/PistonControl.java (hunk 4 above)
Comment (mxgordon) [reply subsystems/PistonControl.java]: you should also make the 2d array, so the for loop doesn't have the `+=2`, cause that causes issues elsewhere

File: subsystems/PistonControl.java (hunk 5 above)
Comment (mxgordon) [reply subsystems/PistonControl.java]: the `+=1` will actually cause `doubleSolenoidPistons[i]` to index too far if there is more than 1 piston

File: subsystems/PistonControl.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        }
+    }
+
+    public void pistonsSet(Value direction) {
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: I like this! However, you should also include `extend()` and `retract()` methods that just call this method with the correct direction

**Review by mxgordon — CHANGES_REQUESTED** [review]

File: subsystems/PistonControl.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package org.frc5587.lib.subsystems;
+
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+
+import static edu.wpi.first.wpilibj.DoubleSolenoid.Value;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class PistonControl extends SubsystemBase{
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: this should also be an abstract class

**Review by mxgordon — CHANGES_REQUESTED** [review]

File: subsystems/PistonControl.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        for(DoubleSolenoid piston : doubleSolenoidPistons) {
+            piston.set(direction);
+        }
+    }
+
+    // Utilize these 2 methods as alternatives to the pistonsSet method
+    public void extend() {
+        for(DoubleSolenoid piston : doubleSolenoidPistons) {
+            piston.set(Value.kForward);
+        }
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: make this call `pistonsSet()`

File: subsystems/PistonControl.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public void retract() {
+        for(DoubleSolenoid piston : doubleSolenoidPistons) {
+            piston.set(Value.kReverse);
+        }
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: Same here

File: subsystems/PistonControl.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+
+import static edu.wpi.first.wpilibj.DoubleSolenoid.Value;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public abstract class PistonControl extends SubsystemBase{
+    private DoubleSolenoid[] doubleSolenoidPistons;
+
+    public PistonControl(int[][] pistonIDS) {
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: Give a brief explanation of the purpose of this class and the parameters. To autopopulate the javadoc, type `/**` and a box should appear saying "java docs" or smth, just hit enter

**Review by mxgordon — CHANGES_REQUESTED** [review]

File: subsystems/PistonControl.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    // Implementation: pistonsSet(kForward)
+    public void pistonsSet(Value direction) {
+        for (DoubleSolenoid piston : doubleSolenoidPistons) {
+            piston.set(direction);
+        }
+    }
+
+    // Utilize these 2 methods as alternatives to the pistonsSet method
+    public void extend() {
+        for (DoubleSolenoid piston : doubleSolenoidPistons) {
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: you don't need the for loop

File: subsystems/PistonControl.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+            pistonsSet(Value.kForward);
+        }
+    }
+
+    public void retract() {
+        for (DoubleSolenoid piston : doubleSolenoidPistons) {
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: or here

File: subsystems/PistonControl.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    // Implementation: pistonsSet(kForward)
+    public void pistonsSet(Value direction) {
+        for (DoubleSolenoid piston : doubleSolenoidPistons) {
+            piston.set(direction);
+        }
+    }
+
+    // Utilize these 2 methods as alternatives to the pistonsSet method
+    public void extend() {
+            pistonsSet(Value.kForward);
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: bruh, fix the indents

File: subsystems/PistonControl.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public void retract() {
+            pistonsSet(Value.kReverse);
```
Comment (mxgordon) [inline subsystems/PistonControl.java]: and here

**Review by mxgordon — CHANGES_REQUESTED** [review]

**Review by mxgordon — APPROVED** [review]

**Pull request merged**